File: solutions/UVA/11786 - Global Raining at Bididibus.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    map<int, vector<int> >::iterator it;
    int n, current, minimum;
    double area1, area2;
    char symbol;
    int numbers[10005];
    map<int, vector<int> > equals;
    string symbols;

    cin >> n;
    
    for(int t=0; t<n; t++) {
        cin >> symbols;
        current = 0;
        minimum = 10005;
        for(int j=0; j<symbols.size(); j++) {
            symbol = symbols[j];
            if(symbol == '/') {
                current = current + 1;
            }
            else if(symbol == '\\') {
                current = current - 1;
            }

            minimum = min(minimum, current);
            numbers[j+1] = current;
        }
        for(int j=1; j<=symbols.size(); j++) {
            numbers[j] = numbers[j] - minimum;
        }

        if(symbols[0] == '/') {
            numbers[0] = numbers[1] - 1;
        }else if(symbols[0] == '\\') {
            numbers[0] = numbers[1] + 1;
        }else{
            numbers[0] = numbers[1];
        }


        for(int i=0; i<=symbols.size(); i++) {
            int number = numbers[i];
            auto iter = equals.find(number);

            if (iter == equals.end() ) {
                equals[number] = vector<int>();
            }
                
            equals[number].push_back(i);
        }

        area2 = 0;
        for(int i=0; i<=symbols.size(); i++) {
            if (numbers[i] != numbers[i+1])
                area2 = area2 + max(numbers[i], numbers[i+1]) - 0.5;
            else
                area2 = area2 + numbers[i];
        }


        for ( it = equals.begin(); it != equals.end(); it++ ) {
            int key = it->first;
            for(int k=0; k<equals[key].size()-1; k++) {
                int i = equals[key][k];
                int j = equals[key][k+1];
    
                int* maximum = max_element(numbers+i, numbers+j);
                if(*maximum <= key) {
                    for(int l=i; l<=j; l++) {
                        numbers[l] = key;
                    }
                }

            }
        }

        area1 = 0;
        for(int i=0; i<=symbols.size(); i++) {
            if (numbers[i] != numbers[i+1])
                area1 = area1 + max(numbers[i], numbers[i+1]) - 0.5;
            else
                area1 = area1 + numbers[i];
        }
    
        int res = area1-area2;
    
        cout << res << endl;
        equals.clear();
    
    }
    return 0;
}

File: solutions/UVA/11786 - Global Raining at Bididibus.py
n = int(input())

for i in range(n):
    symbols = input()
    numbers = []
    current = 0
    minimum = 10005
    for symbol in symbols:
        if symbol == "/":
            current = current + 1
        elif symbol == "\\":
            current = current - 1
        minimum = min(minimum, current)
        numbers.append(current)
    numbers = list(map(lambda x: x - minimum, numbers))

    if symbols[0] == "/":
        numbers.insert(0, numbers[0] - 1)
    elif symbols[0] == "\\":
        numbers.insert(0, numbers[0] + 1)
    else:
        numbers.insert(0, numbers[0])

    equals = {}

    for i, number in enumerate(numbers):
        if number not in equals:
            equals[number] = []
        equals[number].append(i)

    numbers_copy = numbers.copy()

    keys = list(equals.keys())
    keys.sort()

    for key in keys:
        for k in range(len(equals[key]) - 1):
            i = equals[key][k]
            j = equals[key][k+1]

            error = False
            for l in range(i, j+1):
                if(numbers[l] > key):
                    error = True
                    break
            if not error:
                for l in range(i, j+1):
                    numbers[l] = key

    area1, area2 = 0, 0
    for i in range(len(numbers) - 1):
        if numbers[i] != numbers[i+1]:
            area1 = area1 + max(numbers[i], numbers[i+1]) - 0.5
        else:
            area1 = area1 + numbers[i]

        if numbers_copy[i] != numbers_copy[i+1]:
            area2 = area2 + max(numbers_copy[i], numbers_copy[i+1]) - 0.5
        else:
            area2 = area2 + numbers_copy[i]

    res = int(area1-area2)

    print(res)
